packages/blocks/src/page-block/edgeless/connector-manager.ts

```diff
@@ -0,0 +1,944 @@
+    }
+    startPoint = this._conversion(startPoint);
+    endPoint = this._conversion(endPoint);
+    nextStartPoint = this._conversion(nextStartPoint);
+    lastEndPoint = this._conversion(lastEndPoint);
+    const result = this.getConnectablePoints(
+      startPoint,
+      endPoint,
+      nextStartPoint,
+      lastEndPoint,
+      sb,
+      eb,
+      esb,
+      eeb
+    );
+    const points = result.points;
+    nextStartPoint = result.sp;
+    lastEndPoint = result.ep;
+    const finalPoints = this.filterConnectablePoints(
+      this.filterConnectablePoints(points, sb),
+      eb
+    );
+
+    if (!end) {
+      if (
+        Math.abs(endPoint[0] - startPoint[0]) >
+        Math.abs(endPoint[1] - startPoint[1])
+      ) {
+        endPoint[0] += sign(endPoint[0] - startPoint[0]);
+      } else {
+        endPoint[1] += sign(endPoint[1] - startPoint[1]);
+      }
+    }
+    if (!start) {
+      if (
+        Math.abs(endPoint[0] - startPoint[0]) >
+        Math.abs(endPoint[1] - startPoint[1])
+      ) {
+        startPoint[0] -= sign(endPoint[0] - startPoint[0]);
+      } else {
+        startPoint[1] -= sign(endPoint[1] - startPoint[1]);
+      }
+    }
+    this._aStarRunner = new AStarRunner(
+      finalPoints,
+      nextStartPoint,
+      lastEndPoint,
+      startPoint,
+      endPoint,
+      blocks,
+      expandBlocks
+    );
+    this._aStarRunner.run();
+    let path = this._aStarRunner.path;
+    if (!end) path.pop();
+    if (!start) path.shift();
+
+    path = this._mergePath(path);
+    return path;
+  }
+
+  private _mergePath(points: IVec[]) {
+    if (points.length === 0) return [];
+    const rst: IVec[] = [[points[0][0], points[0][1]]];
+    for (let i = 1; i < points.length - 1; i++) {
+      const cur = points[i];
+      const last = points[i - 1];
+      const next = points[i + 1];
+      if (
+        almostEqual(last[0], cur[0], 0.02) &&
+        almostEqual(cur[0], next[0], 0.02)
+      )
+        continue;
+      if (
+        almostEqual(last[1], cur[1], 0.02) &&
+        almostEqual(cur[1], next[1], 0.02)
+      )
+        continue;
+      rst.push([cur[0], cur[1]]);
+    }
+    rst.push(points[points.length - 1]);
+    for (let i = 0; i < rst.length - 1; i++) {
+      const cur = rst[i];
+      const next = rst[i + 1];
+      try {
+        assertEquals(
+          almostEqual(cur[0], next[0], 0.02) ||
+            almostEqual(cur[1], next[1], 0.02),
+          true
+        );
+      } catch (e) {
+        console.log(points);
+        console.log(rst);
+      }
+    }
+    return rst;
+  }
+
+  private _conversion(point: IVec) {
+    return [
+      Number(point[0].toFixed(2)),
+      Number(point[1].toFixed(2)),
+      point[2] ?? 0,
+    ];
+  }
+
+  private filterConnectablePoints(points: IVec[], bound: Bound | null) {
+    return points.filter(point => {
+      if (!bound) return true;
+      return !bound.isPointInBound(point);
+    });
+  }
+
+  private getConnectablePoints(
+    osp: IVec,
```

Comment:
Do not use overly simplified variable naming, or at lease add comment on their naming. we are NOT writing C.

Reply:
This method doesn't use any `this` properties so it should be extracted as standalone utils. We are NOT writing Java.

Reply:
tomorrow, I will continue improve this method